crypto/crypto/src/macros/src/lib.rs

```diff
@@ -36,3 +111,327 @@ pub fn silent_debug(source: TokenStream) -> TokenStream {
+    };
+    for variant in try_iter {
+        let variant_ident = &variant.ident;
+        let variant_arg = &variant
+            .fields
+            .iter()
+            .next()
+            .expect("Unrecognized enum for key types")
+            .ty;
+        try_chain.extend(quote!{
+            .or_else(|_err| #variant_arg::try_from(bytes).and_then(|key| Ok(#name::#variant_ident(key))))
+        })
+    }
+
+    quote! {
+        impl core::convert::TryFrom<&[u8]> for #name {
+            type Error = crypto::CryptoMaterialError;
+            fn try_from(bytes: &[u8]) -> std::result::Result<#name, Self::Error> {
+                #try_chain
+            }
+        }
+    }
+}
+
+fn match_enum_to_bytes(name: &Ident, variants: &DataEnum) -> QuoteTokenStream {
+    // the ValidKey dispatch proper
+    let mut match_arms = quote! {};
+    for variant in variants.variants.iter() {
+        let variant_ident = &variant.ident;
+
+        match_arms.extend(quote! {
+            #name::#variant_ident(key) => key.to_bytes().to_vec(),
+        });
+    }
+    match_arms
+}
+
+fn impl_enum_validkey(name: &Ident, variants: &DataEnum) -> TokenStream {
+    let mut try_from = impl_enum_tryfrom(name, variants);
+
+    let to_bytes_arms = match_enum_to_bytes(name, variants);
+
+    try_from.extend(quote! {
+
+        impl ::crypto::ValidKey for #name {
+            fn to_bytes(&self) -> Vec<u8> {
+                match self {
+                    #to_bytes_arms
+                }
+            }
+        }
+    });
+    try_from.into()
+}
+
+#[proc_macro_derive(PublicKey, attributes(PrivateKeyType))]
+pub fn derive_enum_publickey(input: TokenStream) -> TokenStream {
+    let ast = parse_macro_input!(input as DeriveInput);
+
+    let name = &ast.ident;
+    let private_key_type = get_type_from_attrs(&ast.attrs, "PrivateKeyType");
+    match ast.data {
+        Data::Enum(ref variants) => impl_enum_publickey(name, private_key_type, variants),
+        Data::Struct(_) | Data::Union(_) => {
+            panic!("#[derive(PublicKey)] is only defined for enums")
+        }
+    }
+}
+
+fn get_type_from_attrs(attrs: &[syn::Attribute], attr_name: &str) -> String {
+    let mut associated_key_type = None;
+    for attr in attrs {
+        if let syn::Meta::NameValue(val) = attr.parse_meta().unwrap() {
+            if val
+                .path
+                .get_ident()
+                .expect(&format!("Could not parse {} attribute", attr_name)[..])
+                == attr_name
+            {
+                if let syn::Lit::Str(lit) = &val.lit {
+                    associated_key_type = Some(lit.value());
+                }
+            }
+        }
+    }
+
+    associated_key_type.expect(&format!("Could not find {} attribute", attr_name)[..])
+}
+
+fn impl_enum_publickey(name: &Ident, private_key_type: String, variants: &DataEnum) -> TokenStream {
+    let pkt = syn::parse_str::<syn::Type>(&private_key_type).expect("Misformatted PrivateKeyType");
```

Comment:
Pass the private_key_type in as LitStr instead of String, that way it has a Span attached so that if parsing fails we can put the error in the right place. The parse then will look like:

```rust
let pkt: Type = private_key_type.parse()?;
```